include/pisa/vector.hpp

```diff
@@ -0,0 +1,164 @@
+#pragma once
+
+#include <vector>
+
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/zip.hpp>
+
+namespace pisa {
+
+/// A (potentially) type-safe vector.
+///
+/// It derives from `std::vector<V>` and works essentially like one.
+/// The difference is that you define a key type as well, the outcome
+/// being that if you use a strong type key, you can differentiate between
+/// `Vector<IndexType_1, V>` and `Vector<IndexType_2, V>`.
+template <typename K, typename V = K, typename Allocator = std::allocator<V>>
+class Vector : protected std::vector<V> {
```

Comment:
We absolutely do! Of course, we don't have to use it everywhere, but in some parts when we want to make sure we're type safe, this is great.

For instance, we have different mappings: documents to shards, shards to documents, terms to strings, etc., in the same chunk of code, it's very easy to lose track of what is what. Once you have type safety, you can't make an easy mistake. `std::vector` doesn't give you this.

As for the other containers, `vector` is a specific case. The reason is that here we're targeting the key that is not part of a vector type. But if we use, say, `map` it's already good because the key type is a template param.

So this vector is essentially used when we have a strongly typed integer keys. In other words, we model a total function `f: [0, n) -> V`. Alternatively, you could consider it an optimization of a map for all consecutive values between 0 and n (in `K`) but I think it's clearer to call it a vector (because that's what it in fact is, just with a strongly typed key).

Now, there might be other STL containers that need something like this, but only if we actually need to use them this way. Which we probably don't. Examples: `list` isn't even random access so it doesn't fit our use case of mapping; similarly queues are not used like that; or stacks... So that leaves the lonely vector.

Reply:
I've renamed it `VecMap` for a more descriptive name.